File: src/engine/Shader.java
package engine;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

import static org.lwjgl.opengl.GL20.*;

public class Shader {

	private int shaderID;

	public Attribute[] attributes;
	public Uniform[] uniforms;

	/**
	 * creates a new shader
	 * @param path
	 * path to the shader
	 * @param stype
	 * type of shader
	 */
	public Shader(String path, int stype) {
		String shaderSource = "";

		try {
			BufferedReader reader = new BufferedReader(new FileReader(new File(path)));
			String line;
			ArrayList<Attribute> tempattriblist = new ArrayList<>();
			ArrayList<Uniform> tempuniflist = new ArrayList<>();
			while ((line = reader.readLine()) != null) {
				if (line.startsWith("layout (location")) {
					int startIndex = 15;
					for (char c : line.substring(15).toCharArray()) {
						startIndex++;
						if (c == '=') {
							break;
						}
					}
					String locStr = "";
					for (char c: line.substring(startIndex).toCharArray()) {
						startIndex++;
						if (c == ')')  {
							break;
						}else if (c != ' ') {
							locStr += c;
						}
					}
					startIndex++;
					int loc = Integer.parseInt(locStr);
					String type = "";
					for (char c:line.substring(startIndex).toCharArray()) {
						startIndex++;
						if (c == ' ') {
							break;
						}
					}
					for (char c:line.substring(startIndex).toCharArray()) {
						startIndex++;
						if (c != ' ') {
							type += c;
						}else if (type.length() > 0) {
							break;
						}
					}
					String name = "";
					for (char c:line.substring(startIndex).toCharArray()) {
						if (c != ' ' && c != ';'){
							name += c;
						}
					}
					int vecSize = 0;
					if (type.equals("float")) {
						vecSize = 1;
					}else if (type.startsWith("vec")) {
						vecSize = Integer.parseInt(type.substring(3));
					}else if (type.startsWith("mat")) {
						vecSize = Integer.parseInt(type.substring(3));
					}else {
						throw new IOException("Invalid attribute type in shader");
					}
					tempattriblist.add(new Attribute(vecSize, name, loc));
				}else if (line.startsWith("uniform ")) {//uniform vecn name;
					int startIndex = 8;
					String type = "";
					for (char c: line.substring(startIndex).toCharArray()) {
						startIndex++;
						if (c == ' ') {
							break;
						}
						type += c;
					}
					String name = "";
					for (char c: line.substring(startIndex).toCharArray()) {
						if (c != ' ' && c != ';') {
							name += c;
						}
					}
					int vecSize = 0;
					if (type.equals("float")) {
						vecSize = 1;
						tempuniflist.add(new Uniform(name, vecSize));
					}else if (type.startsWith("vec")) {
						vecSize = Integer.parseInt(type.substring(3));
						tempuniflist.add(new Uniform(name, vecSize));
					}else if (type.startsWith("mat")) {
						vecSize = Integer.parseInt(type.substring(3));
						tempuniflist.add(new Uniform(name, vecSize, true));
					}
				}
				shaderSource += line + "\n";
			}
			attributes = new Attribute[tempattriblist.size()];
			for (int i = 0; i<tempattriblist.size();i++) {
				attributes[i] = tempattriblist.get(i);
			}
			uniforms = new Uniform[tempuniflist.size()];
			for (int i = 0; i<tempuniflist.size();i++) {
				uniforms[i] = tempuniflist.get(i);
			}
		}catch (IOException e) {
			e.printStackTrace();
		}
		shaderID = glCreateShader(stype);
		glShaderSource(shaderID, shaderSource);
		glCompileShader(shaderID);
		if(glGetShaderi(shaderID, GL_COMPILE_STATUS) != 1) {
			throw new RuntimeException("Failed to compile shader: " + path + "! " + glGetShaderInfoLog(shaderID));
		}
	}

	public int getID() {
		return shaderID;
	}

	public void destroy() {
		glDeleteShader(shaderID);
	}
}

File: src/main/kotlin/engine/opengl/shaders/Shader.kt
package engine.opengl.shaders

import engine.getResource
import engine.opengl.GLResource
import org.lwjgl.opengl.GL20.*
import kotlin.collections.ArrayList

@Suppress("ConvertSecondaryConstructorToPrimary", "unused", "MemberVisibilityCanBePrivate")
open class Shader : GLResource {

	open val inAttributes: Array<Attribute>
	open val outAttributes: Array<Attribute>
	open val uniforms: Array<ShaderUniform>

	/**
	 * creates a new shader
	 * @param path path to the shader
	 * @param shaderType type of shader
	 */
	constructor(path : String, shaderType : ShaderType) : super(glCreateShader(shaderType.glID)) {
		var shaderSource = ""
		val reader = getResource(path)
		var line = reader.readLine()
		val tempInAttribList = ArrayList<Attribute>()
		val tempOutAttribList = ArrayList<Attribute>()
		val tempUniformList = ArrayList<ShaderUniform>()
		val constArray = ArrayList<Pair<String, Int>>()

		var lastInAttribPos = -1
		var lastOutAttribPos = -1

		shaderSource += "#version 430 core\nlayout (std430) buffer;\n"

		while (line != null) {
			if (line.startsWith("#")) {
				if (line == "#import <gnoise>") {
					val br = getResource("engine/noise.glsl")
					line = br.readText()
				}
				if (line.startsWith("#import \"")) {
					val br = getResource(line.substring(9, line.length - 1))
					line = br.readText()
				}
			} else if (line.startsWith("in") ||
				(line.startsWith("layout") && line.contains(" in "))) {
				val attrib = Attribute(line, lastInAttribPos, lastOutAttribPos)
				lastInAttribPos = attrib.pos
			} else if (line.startsWith("out") ||
				(line.startsWith("layout") && line.contains(" out "))) {
				val attrib = Attribute(line, lastInAttribPos, lastOutAttribPos)
				lastOutAttribPos = attrib.pos
			} else if (line.startsWith("uniform") ||
				(line.startsWith("layout") && line.contains(" uniform "))) {
				for (const in constArray) {
					if (line.contains(const.first)) {
						line = line.replace(const.first, "${const.second}")
					}
				}
				tempUniformList.add(ShaderUniform(line))
			} else if (line.startsWith("const int")) {
				val name = line.substringBefore('=').split(' ')[2]
				val value = line.substringAfter('=').substringBefore(';').trim().toInt()
				constArray.add(Pair(name, value))
			}

			shaderSource += "$line\n"
			line = reader.readLine()
		}
		inAttributes = tempInAttribList.toTypedArray()
		outAttributes = tempOutAttribList.toTypedArray()
		uniforms = tempUniformList.toTypedArray()
		glShaderSource(id, shaderSource)
		glCompileShader(id)

		checkCompileStatus(path)
	}

	private fun checkCompileStatus(path : String) {
		if (glGetShaderi(id, GL_COMPILE_STATUS) != 1) {
			throw GLShaderCompileException(path, id)
		}
	}

	private constructor() : super(0) {
		inAttributes = emptyArray()
		outAttributes = emptyArray()
		uniforms = emptyArray()
	}

	open val uniformCount : Int get() {return uniforms.size}

	/**
	 * deletes the shader
	 */
	override fun destroy() {
		glDeleteShader(id)
	}

	companion object {
		val defaultShader = Shader()
	}
}

@Suppress("unused")
open class GLShaderCompileException(identifier : String, glID : Int) :
		Exception("Failed to compile shader $identifier!\n${glGetShaderInfoLog(glID)}")